### ISSUE TITLE
[clang] Inaccurate warning for direct-initialization of `bool` to `nullptr`

### ISSUE THREAD

**Issue by Eisenwave** [llvm/llvm-project#64555]
Given the following code (https://godbolt.org/z/YrEnGb35W):
```cpp
bool b(nullptr);
```
clang emits the warning:
```
<source>:1:8: warning: implicit conversion of nullptr constant to 'bool' [-Wnull-conversion]
bool b(nullptr);
     ~ ^~~~~~~
       false
```

However, no conversion should take place here; instead, this is a special case:
> Otherwise, if the initialization is *direct-initialization*, the source type is `std​::​nullptr_t`, and the destination type is `bool`, the initial value of the object being initialized is `false`.

\- [[dcl.init.general] p16.8](https://eel.is/c++draft/dcl.init.general-16.8)

## Suggested Fix
The warning should instead state:
```
direct-initialization of bool to nullptr
```

**Comment by Endilll** [comment]
Confirmed on trunk: https://godbolt.org/z/Wdq48W334
```
<source>:1:8: warning: implicit conversion of nullptr constant to 'bool' [-Wnull-conversion]
    1 | bool b(nullptr);
      |      ~ ^~~~~~~
      |        false
1 warning generated.
```
I agree with your analysis, and it seems to be a false-positive to me.

**Comment by Endilll** [comment]
CC @AaronBallman good first issue?

**Comment by AaronBallman** [comment]
This might not be a good first issue -- the problem is that we diagnose here: https://github.com/llvm/llvm-project/blob/f6ee4e3f5562acfb11f3c36f555e37246aaeefea/clang/lib/Sema/SemaChecking.cpp#L14530

but we don't know whether this conversion is for direct initialization or not. `DiagnoseNullConversion()` is called by `CheckImplicitConversion()` which is called all over the place. So it might be tricky to thread through enough information to know whether to issue the conversion warning or not.
